lib/maths/CBoostedTree.cc

```diff
@@ -1082,11 +1120,76 @@ class CBoostedTree::CImpl final {
         }
         return result;
     }
+    //! Get the bounding box to use for hyperparameter optimisation.
+    TDoubleDoublePrVec hyperparameterBoundingBox(double lambda, double gamma) const {
+        TDoubleDoublePrVec result;
+        if (m_LambdaOverride == boost::none) {
+            result.emplace_back(std::log(lambda / 10.0), std::log(10.0 * lambda));
+        }
+        if (m_GammaOverride == boost::none) {
+            result.emplace_back(std::log(gamma / 10.0), std::log(10.0 * gamma));
+        }
+        if (m_FeatureBagFractionOverride == boost::none) {
+            result.emplace_back(0.2, 0.8);
+        }
+        if (m_ShrinkageFactorOverride == boost::none) {
+            result.emplace_back(0.8, 1.0);
+        }
+        return result;
+    }
+
+    //! Get the hyperparameters which were last used.
+    TVector currentHyperparameters() const {
+
+        TVector result{this->numberHyperparametersToTune()};
+
+        std::size_t i{0};
+        if (m_LambdaOverride == boost::none) {
+            result(i++) = std::log(m_Lambda);
+        }
+        if (m_GammaOverride == boost::none) {
+            result(i++) = std::log(m_Gamma);
+        }
+        if (m_FeatureBagFractionOverride == boost::none) {
+            result(i++) = m_FeatureBagFraction;
+        }
+        if (m_ShrinkageFactorOverride == boost::none) {
+            result(i++) = m_ShrinkageFactor;
+        }
+
+        return result;
+    }
+
+    //! Set the next hyperparameters to use.
+    void nextHyperparameters(TVector parameters) {
+
+        std::size_t i{0};
+        if (m_LambdaOverride == boost::none) {
+            m_Lambda = std::exp(parameters(i++));
+        }
+        if (m_GammaOverride == boost::none) {
+            m_Gamma = std::exp(parameters(i++));
+        }
+        if (m_FeatureBagFractionOverride == boost::none) {
+            m_FeatureBagFraction = parameters(i++);
+        }
+        if (m_ShrinkageFactorOverride == boost::none) {
+            m_ShrinkageFactor = parameters(i++);
+        }
+
+        LOG_TRACE(<< "lambda = " << m_Lambda << ", gamma = " << m_Gamma
+                  << ", feature bag fraction = " << m_FeatureBagFraction
+                  << ", shrinkage = " << m_ShrinkageFactor);
+    }
 
     //! Capture the current hyperparameter values.
-    SHyperparameters captureHyperparameters() {
-        return {*m_Lambda, *m_Gamma, m_FeatureBagFraction, m_ShrinkageFactor,
-                m_FeatureSampleProbabilities};
+    void captureBestHyperparameters(double loss) {
+        if (loss < m_BestForestTestLoss) {
+            m_BestForestTestLoss = loss;
+            m_BestHyperparameters =
+                SHyperparameters{m_Lambda, m_Gamma, m_FeatureBagFraction,
+                                 m_ShrinkageFactor, m_FeatureSampleProbabilities};
```

Comment:
Why isn't this using `currentHyperparameters()`?

Reply:
I think I see the reason why it is not used. `currentHyperparameters()` is the current `tunable` parameters, not the absolute current hyperparameters to be used when building the forest. 

I think it should probably be renamed, it is not obvious that it will not include certain hyperparameters if they are overridden by the user.

Reply:
Yes, I think I agree. I still haven't fully decided what to do with feature weight tuning, although these will be using a different strategy to values returned by `currentHyperparameters()`. I'll try and come up with better naming.

Reply:
I couldn't think of good names for these (bad sign) so I refactored to do the hyperparameter update for a single round into one function; I think this is cleaner. I also factored out cross validation to make the train function as clean as possible. Can you take another look @benwtrent.